### ISSUE TITLE
Jagged array may not be collected in memory captures

### ISSUE THREAD

**Issue by jojo59516** [pschraut/UnityHeapExplorer#9]
Hi
I found jagged array (T[][]) may not be collected in memory captures. 

For codes like
```C#
public class TestComponent : MonoBehaviour
{
    ...
    private TestClass[] array1 =
    {
        new TestClass(),
    };
    private TestClass[][] array2 =
    {
        new [] {new TestClass(), null},
        null,
    };
    ...
}
```
Then managed object `array2` and managed type `TestClass[][]` not found in captured snapshot and managed object of type `TestClass[]` inside `array2` are also not found. (using `Unity 2019.4.5f1`)

However this may be a bug of Unity itself. I have posted a [thread](https://forum.unity.com/threads/memoryprofiler-jagged-array-may-not-be-collected-in-memory-captures.1092970/) in Unity Forum. If It is, would this be a known issue (or something like "Message to Unity Technologies")?

**Comment by pschraut** [comment]
That's interesting. Jagged arrays used to work, they're even part of Heap Explorer's Test Suite. Did you find this issue in a memory capture taken from the editor or a player? Perhaps it makes a difference.

If you capture a memory snapshot in the editor, it can be  an empty project and search for `HeapExplorerWindow` in the C# Objects view, you should find a `m_TestVariables` field in the Inspector. The `m_TestVariables` contain various different types, including 2D and 3D jagged arrays. It does seem to work for me, using Unity 2019.4.20f1, does it work for you?
![jagged_array](https://user-images.githubusercontent.com/6738908/114678373-764eaa00-9d0b-11eb-9f43-123454d98f24.png)

```CSharp
        int[][] m_arrayJaggedInt2D = new int[][]
        {
        new int[] { 0, 1 },
        new int[] { 2, 3, 4 },
        new int[] { 5, 6, 7, 8 },
        new int[] { 9,10,11,12,13 },
        new int[] { 14,15,16,17,18,19 }
        };

**Comment by jojo59516** [comment]
I found this issue in a Windows player.
I will test further later.

**Comment by jojo59516** [comment]
![image](https://user-images.githubusercontent.com/13331846/114693197-f137b500-9d4b-11eb-921b-7e61c66d8346.png)
I tested for Editor. No jagged array is collected :(.
Maybe I can find a fix in one changelog of Unity 2019.4 series.

**Comment by jojo59516** [comment]
Alright, I found it has been fixed in [2019.4.10](https://unity3d.com/unity/whats-new/2019.4.10):
> Profiler: Fixed an issue with multi-dimensional arrays of reference types not having their metadata type being reported (1273416)

**Comment by jojo59516** [comment]
Another bug belongs to Unity is that interned strings cannot be collected.
